src/data/DataStorage.ts

```diff
@@ -0,0 +1,1281 @@
+        }) as DataStorageDimensionInfo);
+
+        const dimensionsIdxMap: Dictionary<number> = {};
+        let prefix = '';
+
+        // Needs to add prefix if key is used in object prototype
+        for (let i = 0; i < dimensions.length; i++) {
+            const name = dimensions[i].name;
+            if ((emptyObj as any)[name] != null) {
+                prefix = '$$';
+            }
+        }
+        for (let i = 0; i < dimensions.length; i++) {
+            const dim = dimensions[i];
+            const name = dim.name;
+            dimensionsIdxMap[prefix + name] = i;
+        }
+
+        // We use different functions because it may be a hotspot code.
+        this.getDimensionIndex = prefix ? function (dim) {
+            return dimensionsIdxMap[prefix + dim];
+        } : function (dim) {
+            return dimensionsIdxMap[dim];
+        };
+
+        this._initDataFromProvider(0, provider.count());
+    }
+
+    getProvider(): DataProvider {
+        return this._provider;
+    }
+
+    getSource(): Source {
+        return this._provider.getSource();
+    }
+
+    getDimensionIndex: (dim: DimensionName) => number;
+
+    getDimensionCount() {
+        return this._dimensions.length;
+    }
+
+    collectOrdinalMeta(
+        dimIdx: number,
+        ordinalMeta: OrdinalMeta
+    ) {
+        const chunk = this._chunks[dimIdx];
+        const dim = this._dimensions[dimIdx];
+        const rawExtents = this._rawExtent;
+
+        const offset = dim.ordinalOffset || 0;
+        const len = chunk.length;
+
+        if (offset === 0) {
+            // We need to reset the rawExtent if collect is from start.
+            // Because this dimension may be guessed as number and calcuating a wrong extent.
+            rawExtents[dimIdx] = getInitialExtent();
+        }
+
+        const dimRawExtent = rawExtents[dimIdx];
+
+        // Parse from previous data offset. len may be changed after appendData
+        for (let i = offset; i < len; i++) {
+            const val = (chunk as any)[i] = ordinalMeta.parseAndCollect(chunk[i]);
+            dimRawExtent[0] = Math.min(val, dimRawExtent[0]);
+            dimRawExtent[1] = Math.max(val, dimRawExtent[1]);
+        }
+
+        dim.ordinalMeta = ordinalMeta;
+        dim.ordinalOffset = len;
+        dim.type = 'ordinal';   // Force to be ordinal
+    }
+
+    getOrdinalMeta(dimIdx: number) {
+        const dimInfo = this._dimensions[dimIdx];
+        const ordinalMeta = dimInfo.ordinalMeta;
+        return ordinalMeta;
+    }
+
+    /**
+     * If we using dataset.
+     * Dimensions types may only know when we initializing series.
+     * So we need to sync the type back to storage when initlializing SeriesData back
+     *
+     * Will return false if dimension type has been known and different from given.
+     * We need to recreate a new data storage in this case.
+     */
+    // TODO Can't sure what's frequency will this validate fail and cause datastorage recreate.
```

Comment:
### Strategy thinking

If a category dimension is used in more than one axis, for example two or more grid layout vertically where there is multiple xAxis but they are totally sync and using the same dimension, the entire dataset will not be shared.

Reply:
I've commented about some changes. Now ordinalMeta should be in the key for query and caching OrdinalMeta. No more validation and recreation should happen when ordinaMeta is not the same.(Still there is a `canUse` check to avoid some unexpected scenarios but it should not exist in our cases.